GoogleSignIn/Sources/GIDEMMSupport.m

```diff
@@ -44,8 +44,55 @@
 // New UIDevice system name for iOS.
 static NSString *const kNewIOSSystemName = @"iOS";
 
+// The error key in the server response.
+static NSString *const kErrorKey = @"error";
+
+// Error strings in the server response.
+static NSString *const kGeneralErrorPrefix = @"emm_";
+static NSString *const kScreenlockRequiredError = @"emm_passcode_required";
+static NSString *const kAppVerificationRequiredErrorPrefix = @"emm_app_verification_required";
+
+// Optional separator between error prefix and the payload.
+static NSString *const kErrorPayloadSeparator = @":";
+
+// A list for recognized error codes.
+typedef enum {
+  ErrorCodeNone = 0,
+  ErrorCodeDeviceNotCompliant,
+  ErrorCodeScreenlockRequired,
+  ErrorCodeAppVerificationRequired,
+} ErrorCode;
+
 @implementation GIDEMMSupport
 
+- (instancetype)init {
+  return [super init];
+}
+
++ (nullable NSError *)handleTokenFetchEMMError:(nullable NSError *)error {
```

Comment:
The original point of this method was just to update the error from the server with the EMM variation supplied by GSI. The other method that you ask about is used in various flows and presents error alerts in some cases.

In any case, the method I added isn't necessary and so I've removed it. Instead, I placed its implementation within `GIDEMMSupport`'s implementation of `-updatedErrorForAuthSession:originalError:` (which it implements to conform to `GTMAuthSessionDelegate`. I think it is more clear now. LMK.

I would like @XiangtianDai to review and verify this. :)

Reply:
Actually, it looks like perhaps my approach is incomplete. In the current implementation, we have:

* [-[GIDAppAuthFetcherAuthorizationWithEMMSupport authorizeRequest:
       completionHandler:]](https://github.com/google/GoogleSignIn-iOS/blob/main/GoogleSignIn/Sources/GIDAppAuthFetcherAuthorizationWithEMMSupport.m#L116)
* Which calls [+[GIDEMMSupport handleTokenFetchError:completion:]](https://github.com/google/GoogleSignIn-iOS/blob/main/GoogleSignIn/Sources/GIDEMMSupport.m#L49)
* Which calls [+[GIDEMMErrorHandler handleErrorFromResponse:completion:]](https://github.com/google/GoogleSignIn-iOS/blob/main/GoogleSignIn/Sources/GIDEMMErrorHandler.m#L59)

`+[GIDEMMErrorHandler handleErrorFromResponse:completion:]` potentially displays some UI, which my approach does not do. My approach does replace the error and send it back up to GTMAppAuth to eventually hand back, but I don't do any UI.

@XiangtianDai please verify if I should be presenting some UI in this case. If so, I will need to find a way to do. It'll be somewhat tricky since the existing mechanism for that uses class methods and block callbacks.

Reply:
Yes, it would be really nice if we can keep the existing behavior of showing the MDM error UI in token fetching that is not part of initial sign-in. If I understand the code correctly this PR shows the MDM error UI in the initial sign-in flow only. This does solve the majority of the use cases, as most apps only sign in once for user identity but don't keep accessing user's data on Google services. However, for these apps that do, and if the device stops being compliant per admin policy after the initial sign-in, without the MDM error UI the user would see access failures without understanding why. Of course, the user can always sign out and sign in again to get the error UI, but still that would be non-ideal user experience. 

Again, not absolutely necessary but really nice to have.

Reply:
Sounds good. We will have to either refactor EMM error handling here or does a minor version release on GTMAppAuth to handle the `GIDGoogleUser` case.

Related, I think [GoogleSignIn-iOS 7.0.0 introduced a bug](https://github.com/google/GoogleSignIn-iOS/issues/304) where the `emmSupport` check will always fail, and will therefore not provide EMM error handling for the case we're discussing here.

We don't currently have any tests for this (though, I am adding something in this PR that will catch future regressions like this); @XiangtianDai are you able to test this somehow? Or, can you recommend a path for me?

Reply:
Yes, I can test the case manually once you're ready, just let me know. We still need unit test for this case though.

Reply:
I believe I have already added a test for this in [-[GIDEMMSupportTest testEMMSupportDelegate]](https://github.com/google/GoogleSignIn-iOS/pull/299/files#diff-d51db5f699811af5b19c97475580c64f1c9bb2dd709c88041740e671b10e44db). Are you thinking of something different?